### PR TITLE
SOL-64939: Added sampler for OTel

### PR DESCRIPTION
**Pull request by TrentDaniel**

### CODE REVIEW

File: howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        self._baggage = None
+
+    def on_message(self, message: 'InboundMessage'):
+        try:
+            tracer = trace.get_tracer(TRACER_NAME)
+            PROPAGATOR = propagate.get_global_textmap()
+             
+            carrier = InboundMessageCarrier(message)
+
+            # inject context form a into the message using a carrier, context is implicit
```
Comment (TamimiGitHub) [inline howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py]: is this comment supposed to say "inject contest from the incoming message into the message"

File: howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+from opentelemetry import propagate
+from opentelemetry import context
+from opentelemetry import trace, baggage
+from opentelemetry.semconv.trace import SpanAttributes, MessagingDestinationKindValues
+from opentelemetry.trace import StatusCode, Status, SpanKind
+from typing import List
+
+from solace_otel.messaging.trace.propagation import InboundMessageCarrier, InboundMessageGetter, OutboundMessageCarrier, OutboundMessageSetter
+
+TRACER_NAME = "my_tracer"
```
Comment (TamimiGitHub) [inline howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py]: can I propose to change the tracer name to `myAppTracer` to follow the same convention we have for the go API sampler

File: howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        return self._baggage
+
+def how_to_publish_with_otel(messaging_service: MessagingService, topic_name: str):
+    """
+    This howto assumes:
+        * that the messaging service has already been created and connected
+        * that a durable exclusive queue has been created and configured to process messages with the given topic
+        * that a telemetry profile has been configured on the broker
+        * that a filter with a subscription to the given topic has been created for that telemetry profile
+        * that these resources are independently cleaned up after this function executes
```
Comment (TamimiGitHub) [inline howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py]: doesnt this howto also assumes a collector is running, whether its the solace cloud hosted one or a local collector?

File: howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py (hunk 2 above)
Comment (TamimiGitHub) [reply howtos/otel/how_to_integrate_opentelemetry_with_pubsubplus_api.py]: https://github.com/SolaceSamples/solace-samples-go/tree/main/patterns/otel-tracing

**Comment by TamimiGitHub** [comment]
Hey @TrentDaniel - thanks for taking a stab at this! I have a couple of comments/questions
1. The python howtos usually have a `@staticmethod def run():` that explicitly calls the howto functions in this sampler. See for example [how_to_publish_with_backpressure.py](https://github.com/SolaceSamples/solace-samples-python/blob/main/howtos/pubsub/how_to_use_publish_with_back_pressure.py)
2. When doing so, and to follow the same conventions across the different API (so far only the [Go API ](https://github.com/SolaceSamples/solace-samples-go/tree/main/patterns/otel-tracing) has samplers for otel), could you please publish on topic `solace/samples/otel-tracing`
3. In the Go API, we had separate samples for publisher and receiver. While I think it can be compact to have a `processor` sample that publishes and receives the message all in one, we will need to add that in the name of the file or as a comment explaining that. What are your thoughts? I'm indifferent about having separate samplers for publisher and subscriber tho I can see having it in one file would be less duplication and better for the flow.
4. There exists a package that enforces otel semantic conventions for Go (https://pkg.go.dev/go.opentelemetry.io/otel@v1.24.0/semconv/v1.17.0) - is there something similar for python that we should be using? I found this https://pypi.org/project/opentelemetry-semantic-conventions/ but i'm not sure if this the conventional or best practice way to use otel with Python. Just a thought

Overall, I did not run this locally with a separate sample but it would be nice to have a runnable method. And if this is just meant to be code snippets on how to use that I would like to propose we have a sample in the [patterns](https://github.com/SolaceSamples/solace-samples-python/tree/main/patterns) that is accompanied with this howto to show a runnable example

**Comment by TamimiGitHub** [comment]
And another question, which might be outside the scope of this PR, when we import packages from the opentel library where and how do we configure the connection to the collector if we want to send application level spans? This code sampler is only involved with configuring the context from the production which will then be propagated through the broker towards the consumer to be extracted

**Comment by TamimiGitHub** [comment]
Pending release of `solace_otel` lib to be published on pypi

**Comment by TrentDaniel** [comment]
> And another question, which might be outside the scope of this PR, when we import packages from the opentel library where and how do we configure the connection to the collector if we want to send application level spans? This code sampler is only involved with configuring the context from the production which will then be propagated through the broker towards the consumer to be extracted

I believe this kind of configuration is indeed outside the scope of this PR. Presumably the configuration would be similar, however I'm not familiar with the otelcollector configuration outside of what I've done for integration testing. In a future commit I'll include a small section which shows how to configure the collector for sending and receiving through SMF, so maybe that will have some of the answers you're looking for.

**Comment by TrentDaniel** [comment]
@TamimiGitHub , just pushed changes in response to PR feedback, let me know if you have any other concerns

**Comment by TrentDaniel** [comment]
@TamimiGitHub , do you have any more change requests?

**Comment by TrentDaniel** [comment]
@TamimiGitHub , This PR has been open for a while, do you have any more change requests??

**Comment by TamimiGitHub** [comment]
Hey @TrentDaniel ! I went over the changes and the last comments, I dont have any other requests! Will merge 👍

**Pull request merged**